File: core/src/main/scala/mybatis/scala/support/objecting/ScalaDefaultObjectFactory.scala
package mybatis.scala.support.objecting

import org.apache.ibatis.reflection.factory.DefaultObjectFactory

import java.util.Properties
import scala.collection.mutable

class ScalaDefaultObjectFactory extends DefaultObjectFactory:

  override def resolveInterface(clazz: Class[?]): Class[?] =
    var resolved = super.resolveInterface(clazz)
    if clazz == classOf[Seq[?]] then resolved = classOf[mutable.ArrayBuffer[?]]
    else if clazz == classOf[Set[?]] then resolved = classOf[mutable.HashSet[?]]
    else if clazz == classOf[Map[?, ?]] then resolved = classOf[mutable.HashMap[?, ?]]
    else resolved = clazz
    resolved

  override def isCollection[T](clazz: Class[T]): Boolean =
    super.isCollection(clazz) ||
      classOf[Seq[?]].isAssignableFrom(clazz) ||
      classOf[Set[?]].isAssignableFrom(clazz)

File: core/src/main/java/mybatis/scala/support/objecting/ScalaDefaultObjectFactory.java
package mybatis.scala.support.objecting;

import org.apache.ibatis.reflection.factory.DefaultObjectFactory;
import scala.collection.mutable.*;

public class ScalaDefaultObjectFactory extends DefaultObjectFactory {

    @Override
    protected Class<?> resolveInterface(Class<?> type) {
        Class<?> resolved = super.resolveInterface(type);

        if (resolved == Seq.class) {
            resolved = ArrayBuffer.class;
        } else if (resolved == Set.class) {
            resolved = HashSet.class;
        } else if (resolved == Map.class) {
            resolved = HashMap.class;
        }

        return resolved;
    }

    @Override
    public <T> boolean isCollection(Class<T> type) {
        boolean collection = super.isCollection(type);
        return collection || Seq.class.isAssignableFrom(type) || Set.class.isAssignableFrom(type);
    }

}
